pymongo/synchronous/client_bulk.py

```diff
@@ -0,0 +1,786 @@
+                database=Database(self.client, "admin"),
+                name="$cmd.bulkWrite",
+            )
+            cmd_cursor = CommandCursor(
+                coll,
+                result["cursor"],
+                conn.address,
+                session=session,
+                explicit_session=session is not None,
+                comment=self.comment,
+            )
+            cmd_cursor._maybe_pin_connection(conn)
+
+            # Iterate the cursor to get individual write results.
+            try:
+                for doc in cmd_cursor:
+                    original_index = doc["idx"] + self.idx_offset
+                    op_type, op = self.ops[original_index]
+
+                    if not doc["ok"]:
+                        result["writeErrors"].append(doc)
+                        if self.ordered:
+                            return
+
+                    # Record individual write result.
+                    if doc["ok"] and self.verbose_results:
+                        if op_type == "insert":
+                            inserted_id = op["document"]["_id"]
+                            res = InsertOneResult(inserted_id)  # type: ignore[assignment]
+                        if op_type in ["update", "replace"]:
+                            op_type = "update"
+                            res = UpdateResult(doc)  # type: ignore[assignment]
+                        if op_type == "delete":
+                            res = DeleteResult(doc)  # type: ignore[assignment]
+                        full_result[f"{op_type}Results"][original_index] = res
+
+            except Exception as exc:
+                # Attempt to close the cursor, then raise top-level error.
+                if cmd_cursor.alive:
+                    cmd_cursor.close()
+                result["error"] = _convert_client_bulk_exception(exc)
+
+    def _execute_command(
+        self,
+        write_concern: WriteConcern,
+        session: Optional[ClientSession],
+        conn: Connection,
+        op_id: int,
+        retryable: bool,
+        full_result: MutableMapping[str, Any],
+        final_write_concern: Optional[WriteConcern] = None,
+    ) -> None:
+        """Internal helper for executing batches of bulkWrite commands."""
+        db_name = "admin"
+        cmd_name = "bulkWrite"
+        listeners = self.client._event_listeners
+
+        # Connection.command validates the session, but we use
+        # Connection.write_command
+        conn.validate_session(self.client, session)
+
+        bwc = self.bulk_ctx_class(
+            db_name,
+            cmd_name,
+            conn,
+            op_id,
+            listeners,  # type: ignore[arg-type]
+            session,
+            self.client.codec_options,
+        )
+
+        while self.idx_offset < self.total_ops:
+            # If this is the last possible batch, use the
+            # final write concern.
+            if self.total_ops - self.idx_offset <= bwc.max_write_batch_size:
```

Comment:
This condition does not represent the final batch, for example if there are 10 InsertOne ops of size 16MB then this condition is true but multiple batches are still required. Could you file a bug ticket for this?